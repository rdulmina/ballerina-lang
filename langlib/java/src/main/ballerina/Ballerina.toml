[package]
org = "ballerina"
name = "java"
version = "0.9.0"
