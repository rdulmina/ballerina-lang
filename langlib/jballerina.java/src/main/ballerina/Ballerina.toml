[package]
org = "ballerina"
name = "jballerina.java"
version = "0.9.0"
