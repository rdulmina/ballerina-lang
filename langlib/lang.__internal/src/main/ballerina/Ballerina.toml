[package]
org = "ballerina"
name = "lang.__internal"
version = "0.1.0"
