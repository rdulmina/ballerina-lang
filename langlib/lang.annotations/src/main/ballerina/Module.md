## Module Overview

This module provides predefined annotations and default error type declaration.
