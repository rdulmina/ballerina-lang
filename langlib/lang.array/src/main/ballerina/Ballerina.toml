[package]
org = "ballerina"
name = "lang.array"
version = "1.1.0"
