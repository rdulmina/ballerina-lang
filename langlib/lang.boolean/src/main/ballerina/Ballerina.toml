[package]
org = "ballerina"
name = "lang.boolean"
version = "1.0.0"
