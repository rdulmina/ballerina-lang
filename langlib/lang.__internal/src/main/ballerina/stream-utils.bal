// Copyright (c) 2020 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
//
// WSO2 Inc. licenses this file to you under the Apache License,
// Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import ballerina/jballerina.java;

# This file contains the utility functions implemented natively which are used by the 'stream' langib module.

# A type parameter that is a subtype of `any|error`.
# Has the special semantic that when used in a declaration
# all uses in the declaration must refer to same type.
@typeParam
public type ErrorType error?;

@typeParam
public type Type any|error;

@typeParam
public type Type1 any|error;

# An `EmptyIterator` which returns nil on next() method invocation.
class EmptyIterator {
    public isolated function next() returns record {|Type value;|}|ErrorType? {
        return ();
    }
}

# Sets the narrowed type of the `value`.
#
# + td - The narrowed type to be set.
# + val - The value of which the type being set.
# + return - The value with the narrowed type.
public isolated function setNarrowType(typedesc<Type> td, record {|Type value;|} val)
    returns record {|Type value;|} = @java:Method {
            'class: "org.ballerinalang.langlib.internal.SetNarrowType",
            name: "setNarrowType"
        } external;

# Takes in an iterator object and returns a new stream out of it.
#
# + td - A type description.
# + iteratorObj - An iterator object.
# + return - New stream containing results of `iteratorObj` object's next function invocations.
public isolated function construct(typedesc<Type> td, object { public isolated function next() returns
        record {|Type value;|}|ErrorType?;} iteratorObj = new EmptyIterator())
        returns stream<Type, ErrorType> = @java:Method {
            'class: "org.ballerinalang.langlib.internal.Construct",
            name: "construct"
        } external;

# Takes a typedesc of an array or stream and returns the typedesc of the element or constraint type.
#
# + td - An array or stream type desc.
# + return - The typedesc of the element, constraint type.
public isolated function getElementType(typedesc<Type[]> | typedesc<stream<Type>> td) returns typedesc<Type> = @java:Method {
    'class: "org.ballerinalang.langlib.internal.GetElementType",
    name: "getElementType"
} external;

# Change the given filter function's parameter to any|error and returns the same function.
#
# + func - The input filter function
# + return - The input function with the changed parameter type
public isolated function getFilterFunc(any func) returns function(Type) returns boolean = @java:Method {
    'class: "org.ballerinalang.langlib.internal.GetFilterFunc",
    name: "getFilterFunc"
} external;

# Change the given map function's parameter to any|error and returns the same function.
#
# + func - The input map function
# + return - The input function with the changed parameter type
public isolated function getMapFunc(any func) returns function(Type) returns Type1 = @java:Method {
    'class: "org.ballerinalang.langlib.internal.GetMapFunc",
    name: "getMapFunc"
} external;

# Get the return type of a given function
#
# + func - The input function
# + return - The typedesc of the return type of the input function
public isolated function getReturnType(any func) returns typedesc<Type> = @java:Method {
    'class: "org.ballerinalang.langlib.internal.GetReturnType",
    name: "getReturnType"
} external;

# Takes in a stream and returns the iterator object of that stream.
#
# + strm - The stream
# + return - An abstract object which is iterable
public isolated function getIteratorObj(stream<Type, ErrorType> strm) returns
    object {
        public isolated function next() returns record {|Type value;|}|ErrorType?;} |
    object {
        public isolated function next() returns record {|Type value;|}|ErrorType?;
        public isolated function close() returns ErrorType?;
    } = @java:Method {
        'class: "org.ballerinalang.langlib.internal.GetIteratorObj",
        name: "getIteratorObj"
    } external;

# Invoke a non-isolated function as an external function (workaround for isolated invocations).
#
# + func - function to invoke
# + args - args for the function to invoke
# + return - result of the invocation
public isolated function invokeAsExternal(any func, any|error... args) returns any|error = @java:Method {
    'class: "org.ballerinalang.langlib.internal.InvokeAsExternal",
    name: "invokeAsExternal"
} external;
