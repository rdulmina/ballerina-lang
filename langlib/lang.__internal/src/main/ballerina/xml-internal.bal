// Copyright (c) 2020 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
//
// WSO2 Inc. licenses this file to you under the Apache License,
// Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import ballerina/java;

# Get elements matching at least one of `elemNames`
#
# + x - The xml source
# + elemNames - ualified name of the elements to filter
# + return - All elements that match elemNames
public function getElements(xml x, string... elemNames) returns xml = @java:Method {
    'class: "org.ballerinalang.langlib.internal.GetElements",
    name: "getElements"
} external;

# Get childElements matching `elemNames` and childIndex for each children sequence.
# xml x = xml `<a><a0>x</a0><a0>y</a0></a><b><a0>j</a0><a0>k</a0></b>`;
# x.getFilteredChildrenFlat(0, "a0") => <a0>x</a0><a0>j</a0>
# x.getFilteredChildrenFlat(-1) => <a0>x</a0><a0>y</a0><a0>j</a0><a0>k</a0>
#
# + x - The xml source
# + index - child index to select from each child sequence, -1 to select all elements
# + elemNames - Qualified name of the elements to filter,
# + return - All child elements matching `index` condition and  `elemNames` condition.
public function getFilteredChildrenFlat(xml x, int index, string... elemNames) returns xml = @java:Method {
    'class: "org.ballerinalang.langlib.internal.GetFilteredChildrenFlat",
    name: "getFilteredChildrenFlat"
} external;

# Searches in children recursively for elements matching the qualified name and returns a sequence containing them
# all. Does not search within a matched result.
#
# + x - The xml source
# + qname - Qualified name of the element
# + return - All the descendants that matches the given qualified name, as a sequence
public function selectDescendants(xml x, string... qname) returns xml = @java:Method {
    'class: "org.ballerinalang.langlib.internal.SelectDescendants",
    name: "selectDescendants"
} external;

# Return attribute matching expanded attribute name
#
# + x - The xml value
# + attributeName - Attribute name in expanded from
# + isOptionalAccess - Is this a optoinal access expression or not
# + return - Attribute value
public function getAttribute(xml x, string attributeName, boolean isOptionalAccess) returns string|error? = @java:Method {
    'class: "org.ballerinalang.langlib.internal.GetAttribute",
    name: "getAttribute"
} external;

# Return name of the element if `x` is a element or nil if element name is not set, else error.
#
# + x - The xml value
# + return - Element name
public function getElementNameNilLifting(xml x) returns string|error? = @java:Method {
    'class: "org.ballerinalang.langlib.internal.GetElementNameNilLifting",
    name: "getElementNameNilLifting"
} external;
