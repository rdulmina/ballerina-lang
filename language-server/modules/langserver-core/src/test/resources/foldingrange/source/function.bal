import ballerina/io;
import ballerina/math;

public function main() {
    io:println("Hello, World!");
}

# Prints PI value
# Refers math library
public function printPI() {
    // Refer symbols of another module.
    // `math:PI` is a qualified identifier. Note the usage of the module prefix.
    float piValue = math:PI;

    // Use the explicit prefix `console` to invoke a function defined in the `ballerina/io` module.
    io:println(piValue);
}
