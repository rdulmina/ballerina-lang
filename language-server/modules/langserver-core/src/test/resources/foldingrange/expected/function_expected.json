{
  "result":[
    {
      "startLine":3,
      "endLine":4,
      "kind":"region"
    },
    {
      "startLine":7,
      "endLine":8,
      "kind":"comment"
    },
    {
      "startLine":9,
      "endLine":15,
      "kind":"region"
    },
    {
      "startLine":0,
      "endLine":1,
      "kind":"region"
    }
  ]
}
